src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -3259,6 +3288,296 @@ protected override void HandleStoreParameter (MethodDefinition method, int index
 				}
 			}
 
+			enum IntrinsicId
+			{
+				None = 0,
+				IntrospectionExtensions_GetTypeInfo,
+				Type_GetTypeFromHandle,
+				Type_MakeGenericType,
+				Type_GetType,
+				Type_GetConstructor,
+				Type_GetMethod,
+				Type_GetField,
+				Type_GetProperty,
+				Type_GetEvent,
+				Expression_Call,
+				Expression_Field,
+				Expression_Property,
+				Expression_New,
+				Activator_CreateInstance_Type,
+				Activator_CreateInstance_AssemblyName_TypeName,
+				Activator_CreateInstanceFrom,
+				Activator_CreateInstanceOfT,
+				AppDomain_CreateInstance,
+				AppDomain_CreateInstanceAndUnwrap,
+				AppDomain_CreateInstanceFrom,
+				AppDomain_CreateInstanceFromAndUnwrap,
+				Assembly_CreateInstance,
+				RuntimeReflectionExtensions_GetRuntimeEvent,
+				RuntimeReflectionExtensions_GetRuntimeField,
+				RuntimeReflectionExtensions_GetRuntimeMethod,
+				RuntimeReflectionExtensions_GetRuntimeProperty
+			}
+
+			static IntrinsicId GetIntrinsicIdForMethod (MethodDefinition calledMethod)
+			{
+				switch (calledMethod.Name) {
+					case "GetTypeInfo" when calledMethod.DeclaringType.Name == "IntrospectionExtensions"
+						&& calledMethod.DeclaringType.Namespace == "System.Reflection":
+						return IntrinsicId.IntrospectionExtensions_GetTypeInfo;
+
+					case "GetTypeFromHandle" when calledMethod.DeclaringType.Name == "Type"
+						&& calledMethod.DeclaringType.Namespace == "System":
+						return IntrinsicId.Type_GetTypeFromHandle;
+
+					case "MakeGenericType" when calledMethod.DeclaringType.Name == "Type"
+						&& calledMethod.DeclaringType.Namespace == "System":
+						return IntrinsicId.Type_MakeGenericType;
+
+					//
+					// System.Reflection.RuntimeReflectionExtensions
+					//
+					// static GetRuntimeEvent (this Type type, string name)
+					// static GetRuntimeField (this Type type, string name)
+					// static GetRuntimeMethod (this Type type, string name, Type[] parameters)
+					// static GetRuntimeProperty (this Type type, string name)
+					//
+					case var getRuntimeMember when (getRuntimeMember.StartsWith ("GetRuntime")
```

Comment:
Later (maybe not this pull request): making this a normal switch over a bunch of strings would get the C# compiler to emit a hashtable lookup. Switch statements with complex `when` conditions have perf characteristics that are hard for me to reason about (it's possible a lot of this is going to be cascading `if` checks). We'll need this method to be as fast as possible.

Reply:
I agree, I just didn't want to rewrite it too much yet.

Reply:
I wonder if there would be any advantage to refactoring this to an annotation stage step that populates a table in `AnnotationStore`.  The pros being 

* We could easily parallelize parts of the annotating process
* The annotation stage (or maybe it's own step) might be a nicer "choke point" as you talked about in a another comment?

Downside being, you end up processing methods that are never called.

Reply:
I was actually thinking about adding a cache similar to what you suggest. But instead of filling it by scanning everything (a separate step or similar) we could lookup the intrinsic. There's not that many of them and we even know which assemblies they should live in. That would avoid scanning all methods, would be much faster and would let us have only "positive" cache (only the intrinsic methods would be stored in the cache).